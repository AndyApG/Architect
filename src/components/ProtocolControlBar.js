import React, { useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import history from '@app/history';
import { UnsavedChanges } from '@components/Dialogs';
import { Button, Spinner } from '@codaco/ui';
import { getProtocol } from '@selectors/protocol';
import { getHasUnsavedChanges } from '@selectors/session';
import { actionCreators as dialogActions } from '@modules/dialogs';
import { actionCreators as userActions, actionLocks as protocolsLocks } from '@modules/userActions';
import { selectors as statusSelectors } from '@modules/ui/status';
import logoutIcon from '@app/images/home/log-out.svg';
import ControlBar from '@components/ControlBar';

const unsavedChangesDialog = UnsavedChanges({
  message: (
    <div>
      Are you sure you want to go back to the start screen?
      <p><strong>Unsaved changes will be lost!</strong></p>
    </div>
  ),
  confirmLabel: 'Go to start screen',
});

const ProtocolControlBar = () => {
  const dispatch = useDispatch();

  const hasUnsavedChanges = useSelector(state => getHasUnsavedChanges(state));
  const hasAnyStages = useSelector(state => getProtocol(state).stages.length > 0);
  const isSaving = useSelector(state => statusSelectors.getIsBusy(state, protocolsLocks.saving));
  const showSaveButton = hasAnyStages && hasUnsavedChanges;

  const saveNetcanvas = useCallback(() => dispatch(userActions.saveNetcanvas()), [dispatch]);

  const handleClickStart = useCallback(
    () => Promise.resolve()
      .then(() => {
        if (!hasUnsavedChanges) { return true; }

        return dispatch(dialogActions.openDialog(unsavedChangesDialog));
      })
      .then((confirm) => {
        if (!confirm) { return; }
        history.push('/');
      }),
    [dispatch, hasUnsavedChanges],
  );

  return (
    <ControlBar
      show
      secondaryButtons={[
        <Button
          key="return-button"
          icon={<div><img src={logoutIcon} alt="Return to start screen" /></div>}
          color="platinum"
          onClick={handleClickStart}
        >
          Return to start screen
        </Button>,
      ]}
      buttons={[
        ...(showSaveButton ? [<Button
          key="save-button"
          onClick={saveNetcanvas}
          color="primary"
          data-variant="save"
          disabled={isSaving}
          content={isSaving ? 'Saving...' : 'Save Changes'}
          iconPosition="right"
          icon={isSaving ? <div><Spinner size="0.5rem" /></div> : 'arrow-right'}
        />] : []),
      ]}
    />
  );
};

export default ProtocolControlBar;
