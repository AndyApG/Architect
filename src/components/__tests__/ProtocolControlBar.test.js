/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import { ProtocolControlBar } from '../ProtocolControlBar';

const mockProps = {
  hasUnsavedChanges: false,
  hasAnyStages: false,
  saveProtocol: () => {},
};

describe('<ProtocolControlBar />', () => {
  it('can render', () => {
    const component = shallow(<ProtocolControlBar {...mockProps} />);

    expect(component).toMatchSnapshot();
  });

  it('Save button is disabled if stages are empty', () => {
    expect(
      shallow(<ProtocolControlBar {...mockProps} hasAnyStages={false} />)
        .find('Button')
        .prop('disabled'),
    ).toBe(true);

    expect(
      shallow(<ProtocolControlBar {...mockProps} hasAnyStages />)
        .find('Button')
        .prop('disabled'),
    ).toBe(false);
  });

  it('Shows protocol bar if there are unsaved changes', () => {
    expect(
      shallow(<ProtocolControlBar {...mockProps} hasUnsavedChanges={false} />)
        .prop('show'),
    ).toBe(false);

    expect(
      shallow(<ProtocolControlBar {...mockProps} hasUnsavedChanges />)
        .prop('show'),
    ).toBe(true);
  });
});
