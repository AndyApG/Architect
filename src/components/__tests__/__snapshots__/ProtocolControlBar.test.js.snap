// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProtocolControlBar /> can render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProtocolControlBar
    hasAnyStages={false}
    hasUnsavedChanges={false}
    saveProtocol={[Function]}
    show={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "buttons": <Button
        color="white"
        content=""
        disabled={true}
        icon={
          <Icon
            className=""
            color=""
            name="arrow-right"
            style={Object {}}
          />
        }
        iconPosition="right"
        onClick={[Function]}
        size=""
      >
        Save
      </Button>,
      "className": "",
      "flip": false,
      "secondaryButtons": null,
      "show": false,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "buttons": <Button
          color="white"
          content=""
          disabled={true}
          icon={
            <Icon
              className=""
              color=""
              name="arrow-right"
              style={Object {}}
            />
          }
          iconPosition="right"
          onClick={[Function]}
          size=""
        >
          Save
        </Button>,
        "className": "",
        "flip": false,
        "secondaryButtons": null,
        "show": false,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
